server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1845,6 +1846,12 @@ final boolean refresh(String source, SearcherScope scope, boolean block) throws
                     } else {
                         refreshed = referenceManager.maybeRefresh();
                     }
+                    final ElasticsearchDirectoryReader current = referenceManager.acquire();
+                    try {
+                        segmentGeneration = current.getIndexCommit().getGeneration();
```

Comment:
Hmm does this make sense if `refreshed == false`? I.e. we used the non-blocking call and it didn't do anything. Can we just use `RefreshResult.UNKNOWN_GENERATION` in that case?